chromadb/test/ingest/test_producer_consumer.py

```diff
@@ -0,0 +1,259 @@
+import pytest
+from itertools import count
+from typing import (
+    Generator,
+    List,
+    Callable,
+    Optional,
+    Dict,
+    Union,
+    Iterator,
+    Sequence,
+    Tuple,
+)
+from chromadb.ingest import Producer, Consumer
+from chromadb.db.impl.sqlite import SqliteDB
+from chromadb.types import (
+    SubmitEmbeddingRecord,
+    Operation,
+    EmbeddingRecord,
+    ScalarEncoding,
+)
+from chromadb.config import System, Settings
+from pytest import FixtureRequest, approx
+from asyncio import Event, wait_for
+from asyncio.exceptions import TimeoutError
+
+
+def sqlite() -> Generator[Tuple[Producer, Consumer], None, None]:
+    """Fixture generator for sqlite Producer + Consumer"""
+    db = SqliteDB(System(Settings(sqlite_database=":memory:", allow_reset=True)))
+    db.start()
+    yield db, db
+    db.stop()
+
+
+def fixtures() -> List[Callable[[], Generator[Tuple[Producer, Consumer], None, None]]]:
+    return [sqlite]
+
+
+@pytest.fixture(scope="module", params=fixtures())
+def producer_consumer(
+    request: FixtureRequest,
+) -> Generator[Tuple[Producer, Consumer], None, None]:
+    yield next(request.param())
+
+
+@pytest.fixture(scope="module")
+def sample_embeddings() -> Iterator[SubmitEmbeddingRecord]:
+    def create_record(i: int) -> SubmitEmbeddingRecord:
+        vector = [i + i * 0.1, i + 1 + i * 0.1]
+        metadata: Optional[Dict[str, Union[str, int, float]]]
+        if i % 2 == 0:
+            metadata = None
+        else:
+            metadata = {"str_key": f"value_{i}", "int_key": i, "float_key": i + i * 0.1}
+
+        record = SubmitEmbeddingRecord(
+            id=f"embedding_{i}",
+            embedding=vector,
+            encoding=ScalarEncoding.FLOAT32,
+            metadata=metadata,
+            operation=Operation.ADD,
+        )
+        return record
+
+    return (create_record(i) for i in count())
+
+
+class CapturingConsumeFn:
+    embeddings: List[EmbeddingRecord]
+    waiters: List[Tuple[int, Event]]
+
+    def __init__(self) -> None:
+        self.embeddings = []
+        self.waiters = []
+
+    def __call__(self, embeddings: Sequence[EmbeddingRecord]) -> None:
+        self.embeddings.extend(embeddings)
+        for n, event in self.waiters:
+            if len(self.embeddings) >= n:
+                event.set()
+
+    async def get(self, n: int) -> Sequence[EmbeddingRecord]:
+        "Wait until at least N embeddings are available, then return all embeddings"
+        if len(self.embeddings) >= n:
+            return self.embeddings[:n]
+        else:
+            event = Event()
+            self.waiters.append((n, event))
+            # timeout so we don't hang forever on failure
+            await wait_for(event.wait(), 10)
```

Comment:
should this timeout be parameterized somehow? Can we leave a TODO for that if so.